### PR TITLE
Fix order of ranges in MergeTreeDataSelectExecutor

### LINKED ISSUE #8964: DB::Exception: RangeReader read 7523 rows, but 7550 expected..
Query:

```
SELECT 
         *
     FROM reporting_events
     WHERE (device_id = 51);
```

Result:

`DB::Exception: RangeReader read 7523 rows, but 7550 expected..`

Table:

```
CREATE TABLE reporting_events (
  id UInt64,
  a UInt32,
  b UInt64,
  c UInt32,
  d DEFAULT 0,
  e UInt32 DEFAULT 0,
  device_id UInt32,
  ts DateTime DEFAULT now(),
  name String,
  type Enum8('ONLINE'      = 0,
             'OFFLINE'     = 1,
             'INFORMATION' = 2,
             'WARNING'     = 3,
             'CRITICAL'    = 4),
  description String
)
ENGINE = MergeTree()
PARTITION BY toYYYYMM(ts)
ORDER BY (a, b, e, device_id, ts, type)
TTL ts + INTERVAL 6 MONTH;
```

Clickhouse version 20.1.2, same with 20.1.3.7. Probably this error due to update to the latest clickhouse version.

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix order of ranges while reading from` MergeTree` table in one thread. It could lead to exceptions from `MergeTreeRangeReader` or wrong query results.

Fixes #8964.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Actually the mistake was that [here](https://github.com/ClickHouse/ClickHouse/blob/016afd6c0edc4201ef2d1c752a5ad1e10ffb3cef/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L793-L798) ranges were passed to MergeTreeSelectProcessor in reversed order, while it expected them in direct order. Before ranges were reversed [here](https://github.com/ClickHouse/ClickHouse/blob/016afd6c0edc4201ef2d1c752a5ad1e10ffb3cef/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L722).

File: dbms/src/Storages/MergeTree/MergeTreeReadPool.cpp [hunk 1]
```diff
@@ -251,6 +240,10 @@ void MergeTreeReadPool::fillPerThreadInfo(
 {
     threads_tasks.resize(threads);
 
+    /// Let the ranges be listed from right to left so that the leftmost range can be dropped using `pop_back()`.
```
Comment (akuzm) [inline dbms/src/Storages/MergeTree/MergeTreeReadPool.cpp]: If the only reason to reverse the array is to have O(1) deletion, we can just use `std::deque`.

Comment (CurtizJ) [reply dbms/src/Storages/MergeTree/MergeTreeReadPool.cpp]: Yes, we can. Actually I don't why originally std::vector was used.

**Comment by alesapin** [comment]
> Actually the mistake was that [here](https://github.com/ClickHouse/ClickHouse/blob/016afd6c0edc4201ef2d1c752a5ad1e10ffb3cef/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L793-L798) ranges were passed to MergeTreeSelectProcessor in reversed order, while it expected them in direct order. Previosly ranges were reversed [here](https://github.com/ClickHouse/ClickHouse/blob/016afd6c0edc4201ef2d1c752a5ad1e10ffb3cef/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp#L722).

Seems like single-threaded read order was broken here too.
https://github.com/ClickHouse/ClickHouse/pull/9048

**Comment by CurtizJ** [comment]
[This](https://github.com/ClickHouse/ClickHouse/blob/2734347cc0c679fb979f5754cf8fe11b67790ad8/dbms/src/Storages/MergeTree/MergeTreeThreadSelectBlockInputProcessor.cpp#L69) and [this](https://github.com/ClickHouse/ClickHouse/blob/9251fb803afeb8109aaa612f4d5252f1d787fd50/dbms/src/Storages/MergeTree/MergeTreeThreadSelectBlockInputProcessor.cpp#L90) lines also looks like bugs, because we want to get rest of the marks starting from the left most, but pass the right most mark as them were reversed. Now, it's ok without changing these lines.

**Review by alesapin — APPROVED** [review]

**Pull request merged**